Add post query cycle script execution hook

agrawalreetika opened the pull request:
This PR includes 2 changes - 
- Add post-query cycle script execution hook
- Add pre stage script execution hook
- Add cache cleanup scripts for Presto coordinator and workers (General Util functions to invalidate different types of worker and coordinator caches

ethanyzhang commented on stage/stage.go:
```diff
@@ -44,6 +44,8 @@ type Stage struct {
 	// Run shell scripts after executing each query.
 	PostQueryShellScripts []string `json:"post_query_scripts,omitempty"`
+	// Run shell scripts after finishing full query cycle runs each query.
+	PostQueryCycleShellScripts []string `json:"post_query_cycle_scripts,omitempty"`
```
@steveburnett for doc

ethanyzhang replied:
A full cycle here means when we set `cold_runs` and `warm_runs`, each query in the benchmark will be run `cold_runs + warm_runs` times in total.

`post_query_scripts` will be called after each query execution, `post_query_cycle_scripts` will be called after all the `cold_runs` and `warm_runs` are done for a unique query.

ethanyzhang commented on benchmarks/test/my_post_query_cycle_script.py:
```diff
@@ -0,0 +1,11 @@
+        sys.exit(-1)
+
+    file_path = sys.argv[1]
+    increment_file_value(file_path)
```
if we touched this file in this script, then I guess the unit test assert result should be updated?

ethanyzhang commented on benchmarks/scripts/cache_cleaning_coordinator_post_query.py:
```diff
@@ -0,0 +1,75 @@
+    conn = create_connection(hostname, username, password, catalogName)
+    cur = conn.cursor()
+    cur.execute(query)
+    rows = cur.fetchall()
```
What is the error behavior?

agrawalreetika commented:
@ethanyzhang Thanks for your review. I have addressed your earlier comments.

Also, I have added one more commit in this PR for adding one more hook to enable `pre-stage script execution hook`. 
Please let me know if that needs to be in a separate PR or If It's Okay to have it in this same PR for review.

yingsu00 commented on benchmarks/scripts/cache_cleaning_workers_post_query.py:
```diff
@@ -0,0 +1,135 @@
+import paramiko
+import argparse
+
+def create_connection(host_name, user_name, user_password, db_name):
```
@ethanyzhang how does pbench manage the utility functions like create_connection? It also appears in benchmarks/scripts/cache_cleaning_coordinator_post_query.py

ethanyzhang replied:
@agrawalreetika I think it may be better to move `create_connection` to a different file then include it in both the worker script and the coordinator script.

Say this file is called `presto_utils.py`, then here you can do:
```python
from presto_utils import create_connection
```

yingsu00 commented on benchmarks/test/my_post_query_cycle_script.py:
```diff
@@ -0,0 +1,11 @@
+import sys
```
This comment is for the two .py file names. Why did you add `my_` in front? I think it's better to remove it

ethanyzhang replied:
this folder is for testing and demoing.

yingsu00 replied:
What does "my" stand for?

ethanyzhang replied:
No meaning. it is for pbench test, not for production.

yingsu00 commented on benchmarks/scripts/cache_cleaning_coordinator_post_query.py:
```diff
@@ -0,0 +1,80 @@
+import prestodb
```
The cache cleaning should be done **before** (not after) each query cycle (cold+warm runs). And it should only be applied to TPC Power Tests.

agrawalreetika replied:
So in the current implementation, I added 2 hooks -
- post_query_scripts: run this script before this stage is started
- pre_stage_scripts  run this script after each query in this stage is complete
Which will eventually trigger the clean-up script, once when the overall query execution starts as part `pre_stage_scripts` and then using `post_query_scripts` after each query cycle (cold+warm runs) the same clean-up script would be executed. So by using both the hooks, I was running clean up script.

But now with this script name, I think it looks a little confusing. I will add one more hook named `pre_query_scripts` just call that before each query cycle (cold+warm runs) for each query and rename these clean up script from `post*` to `pre*`

yingsu00 replied:
Hi @agrawalreetika I think the clean cache scripts should be called in every pre_query_cycle_scripts, not post_query_scripts and pre_stage_scripts. So we want this order suppose cold runs=1 and warm runs=1
clean cache, q1, q1, clean cache, q2, q2, clean cache, q3, q3, ...

ethanyzhang commented on benchmarks/scripts/cache_cleaning_coordinator_post_query.py:
```diff
@@ -0,0 +1,80 @@
+            if rows[0][0] == True:
+                print("Directory list cache clean up is successfull for", catalogName)
+            else:
+                print("Directory list cache clean up is failed for", catalogName)
```
See my comments below at `stage.go` for adding error propagation here.

ethanyzhang commented on benchmarks/scripts/cache_cleaning_coordinator_post_query.py:
```diff
@@ -0,0 +1,80 @@
+            if rows[0][0] == True:
+                print("Metastore cache clean up is successfull for", catalogName)
+            else:
+                print("Metastore cache clean up is failed for", catalogName)
```
Same for error propagation

ethanyzhang commented on benchmarks/scripts/cache_cleaning_workers_post_query.py:
```diff
@@ -0,0 +1,135 @@
+        cleanup_worker_disk_cache(worker_public_ips, native_cache_directory_worker, "centos", args.sshkey)
+
+    if is_worker_os_cache_cleanup_enabled:
+        cleanup_worker_os_cache(worker_public_ips, "centos", args.sshkey)
```
Similar question about propagating errors that could happen here back to pbench. The benchmark should fail if we actually couldn't effectively clean up the cache.

ethanyzhang reviewed (changes requested):
Hi @agrawalreetika, I developed some questions about error propagation. See my comments and I am happy to discuss further.

agrawalreetika commented:
@ethanyzhang @yingsu00  Thanks for your review. I have made the changes based on the review comments. Please check.

yingsu00 commented on benchmarks/scripts/cache_cleaning_coordinator.py:
```diff
@@ -0,0 +1,57 @@
+    # Directory list cache clean up
+    if is_list_cache_cleanup_enabled:
+        for catalogName in catalog_list:
+            print("Cleaning up directory list cache for", catalogName)
```
Should there be a space after "for"? Same for other occurrances

yingsu00 commented on benchmarks/scripts/cache_cleaning_coordinator.py:
```diff
@@ -0,0 +1,57 @@
+        for catalogName in catalog_list:
+            print("Cleaning up directory list cache for", catalogName)
+            rows = clean_directory_list_cache(args.host, args.username, args.password, catalogName)
+            print("directory_list_cache_cleanup_query Query Result:", rows)
```
space after "Result:"?

yingsu00 commented on benchmarks/scripts/cache_cleaning_workers.py:
```diff
@@ -0,0 +1,114 @@
+    is_worker_os_cache_cleanup_enabled = True
+
+    if is_worker_disk_cache_cleanup_enabled:
+        native_cache_directory_worker = "/home/centos/presto/async_data_cache"
```
Would this work on Ubuntu?

agrawalreetika replied:
Currenlty pbench is getting called on Presto clusters, which runs on CentOs, where this is tested.

yingsu00 commented on benchmarks/scripts/cache_cleaning_workers.py:
```diff
@@ -0,0 +1,114 @@
+from mysql_utils import create_connection
```
Where is this being called?

agrawalreetika replied:
In https://github.com/prestodb/pbench/pull/32/files#diff-178b21505ceae58580985f1d1c6c27f05c66411c197c31a5efd7ca0acc94dd32R74 to get the worker list from mysql

yingsu00 commented on stage/stage.go:
```diff
@@ -438,6 +446,11 @@ func (s *Stage) runQueries(ctx context.Context, queries []string, queryFile *str
+		// run post query cycle shell scripts
+		postQueryCycleErr := s.runShellScripts(ctx, s.PostQueryCycleShellScripts)
+		if retErr == nil {
+			retErr = postQueryCycleErr
```
@ethanyzhang @agrawalreetika As we haven't tried it in real runs yet, this may or may not happen. But assume it happens because of some non-recoverable errors, the users won't be able to run any benchmark at all. Maybe we can consider issuing a warning but still continue the queries?

ethanyzhang replied:
There is not a mechanism to preserve the outputs from those scripts. the warning won't be reported. So better to stop.

ethanyzhang replied:
Do we see all benchmark runs will run this script? I think probably we need two modes: with and without this script.
For benchmarks run with this script, we need to make sure the script runs successfully

agrawalreetika commented on benchmarks/scripts/cache_cleaning_coordinator.py:
```diff
@@ -0,0 +1,47 @@
+
+    # Directory list cache clean up
+    if is_list_cache_cleanup_enabled:
+        for cluster_name in catalog_list:
```
Ohh, this was a typo. Fixed it

ethanyzhang commented:
@steveburnett can you confirm completion of doc?

steveburnett commented:
> @steveburnett can you confirm completion of doc?

I will work on the doc for this today, and I will confirm when I am done.

steveburnett commented:
Yesterday I wrote new entries in [Parameters](https://github.com/prestodb/pbench/wiki/Parameters) for 
[post_query_cycle_scripts](https://github.com/prestodb/pbench/wiki/Parameters#post_query_cycle_scripts) and 
[post_query_scripts](https://github.com/prestodb/pbench/wiki/Parameters#post_query_scripts). 

I also added a definition of a cycle to the existing entries for [cold_runs](https://github.com/prestodb/pbench/wiki/Parameters#cold_runs) and [warm_runs](https://github.com/prestodb/pbench/wiki/Parameters#warm_runs).

I'm happy to revise or add more doc for anything that I overlooked, or that could be improved. Let me know!

ethanyzhang commented:
Thank you so much Steve!

agrawalreetika commented:
Hi @steveburnett 
So as part of this change, we have added 3 new parameter options. You can find the names of these in here https://github.com/prestodb/pbench/pull/32/files#diff-0fb84966b346f3820412f8931d2d4e96c9f8e20ca4abb00fecf9c91d39b4a54d

Could you please add all 3 of them in the document?

steveburnett commented:
> Hi @steveburnett So as part of this change, we have added 3 new parameter options. You can find the names of these in here https://github.com/prestodb/pbench/pull/32/files#diff-0fb84966b346f3820412f8931d2d4e96c9f8e20ca4abb00fecf9c91d39b4a54d
> 
> Could you please add all 3 of them in the document?

Yes, I can write entries for pre_stage_scripts, post_stage_scripts, and pre_query_cycle_scripts today. 

Thanks for letting me know! I will comment here when I have added these to the doc.

steveburnett commented:
I have added entries for [pre_stage_scripts](https://github.com/prestodb/pbench/wiki/Parameters#pre_stage_scripts), [post_stage_scripts](https://github.com/prestodb/pbench/wiki/Parameters#post_stage_scripts), and [pre_query_cycle_scripts](https://github.com/prestodb/pbench/wiki/Parameters#pre_query_cycle_scripts) to the [Parameters](https://github.com/prestodb/pbench/wiki/Parameters) page of the PBench documentation.

Let me know if I can help with anything else!
